<center>
    
# 人工智能高阶人才培训班
## 课程作业（二）

---

> 💥说明💥
> 1. 本次作业最晚提交时间： **<font color='red'>2019年6月12日</font>**
> 2. 作业中以`Q-n`标注的几个问题的答案请通过[答题页面](https://app.askform.cn/43245470-5adc-4293-9427-0a33e9743625.aspx?Type=2)连接的Form表格提交，提交中填写的姓名请与本人姓名保持一致；
> 3. 以`Lesson02-HW-{你的名字英文}.ipynb`命名的作业文件请参考[文档](../How-To.md)说明提交到Git的代码仓库；
> 4. 如果完成作业过程中，你有任何新的想法和解题思路，可以完成一篇分享文档(建议用MarkDown撰写)，将根据质量给予⭐奖励
> 5. 🚫 **注意：** 请不要直接编辑该模板文件，建议复制该模板文件生成新的以你自己姓名命名的作业文件（`Lesson02-HW-{你的名字中文}.ipynb`）

### Q1
参考课程中使用`numpy`进行数理统计的过程，使用numpy提供的库`random.normal`生成1万个样本

本题目中使用下表格中**华中地区的女性身高数据**作为参考源

<center>

##### 中国成年人身高，体重均值和标准差的参考值 （2015）
|性别|东北华北|西北|东南|华中|华南|西南|
|-|-|-|-|-|-|-|
|身高，体重|均值，标准差|均值，标准差|均值，标准差|均值，标准差|均值，标准差|均值，标准差|
|男(mm)|1693, 56.6|1684, 53.7|1686, 55,2|1669, 56.3|1650, 57.1|1647, 56.7|
|女(mm)|1586, 51.8|1575, 51.9|1575, 50.8|1560, 50.7|1549, 49.7|1546, 53.9|
|男(kg)|64, 8.2|60, 7.6|59, 7.7|57, 6.9|56, 6.9|55, 6.8|
|女(kg)|55, 7.1|52, 7.1|51, 7.2|50, 6.8|49, 6.5|50, 6.9|


> **注意：**为保证结果的一致性，请每次在你执行随机分布生成之前，使用`np.random.seed(42)`，产生同样的种子值

#### 问： 你生成的分布的实际产生的方差、均值、中位数各是多少(保留两位小数)？
- 50.70, 1560.01, 1575.91
- 50.87, 1560, 1560.69
- 2588.06, 1559.89, 1559.87$\surd$
- 50.87, 1559.89, 1559.87

In [33]:
import numpy as np
np.random.seed(42)
#my code
random_data = np.random.normal(loc=1560,scale=50.7,size=10000)
'Variance:''%.2f' % (np.var(random_data))
'Mean:''%.2f' % (np.mean(random_data))
'Median:''%.2f' % (np.median(random_data))

'Variance:2588.06'

'Mean:1559.89'

'Median:1559.87'

## Q2-Q9
Q2-Q9将使用kaggle入门中常使用的泰坦尼克数据集，详情可以访问[Titanic: Machine Learning from Disaster](https://www.kaggle.com/c/titanic)，为了方便处理我们已经对原始数据进行了初步的处理（如缺失值、异常值等）。

在这一系列题目中，我们将综合使用`numpy`, `pandas`和`matplotlib`解决一个泰坦尼克号数据的初步分析问题。在机器学习中，通过多种不同的方式对数据进行分析有助于及时捕获数据特征和数据之间的关联。

In [86]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt


首先我们使用pandas导入数据

In [87]:
data = pd.read_csv('./data/titanic_train.csv')

查看数据结构和基本组成

In [88]:
data.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

每个字段的含义如下：

| **Variable** | **Definition**                             | **Key**                                        |
| ------------ | ------------------------------------------ | ---------------------------------------------- |
| survival     | Survival                                   | 0 = No, 1 = Yes                                |
| pclass       | Ticket class                               | 1 = 1st, 2 = 2nd, 3 = 3rd                      |
| sex          | Sex                                        |                                                |
| Age          | Age in years                               |                                                |
| sibsp        | # of siblings / spouses aboard the Titanic |                                                |
| parch        | # of parents / children aboard the Titanic |                                                |
| ticket       | Ticket number                              |                                                |
| fare         | Passenger fare                             |                                                |
| cabin        | Cabin number                               |                                                |
| embarked     | Port of Embarkation                        | C = Cherbourg, Q = Queenstown, S = Southampton |


In [89]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

为了让大家快速熟悉pandas，我们准备了几个简单的操作：

1. 让我们尝试从乘客中过滤只显示通过Cherbourg(Embarked=C) 登陆并且付费大于200磅（fare > 200）：

In [90]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)].head()

PassengerId  Survived  Pclass  \
118          119         0       1   
258          259         1       1   
299          300         1       1   
311          312         1       1   
377          378         0       1   

                                                Name     Sex   Age  SibSp  \
118                         Baxter, Mr. Quigg Edmond    male  24.0      0   
258                                 Ward, Miss. Anna  female  35.0      0   
299  Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0      0   
311                       Ryerson, Miss. Emily Borie  female  18.0      2   
377                        Widener, Mr. Harry Elkins    male  27.0      0   

     Parch    Ticket      Fare            Cabin Embarked  
118      1  PC 17558  247.5208          B58 B60        C  
258      0  PC 17755  512.3292              NaN        C  
299      1  PC 17558  247.5208          B58 B60        C  
311      2  PC 17608  262.3750  B57 B59 B63 B66        C  
377      2    113503  211.5000              C82        C

2. 我们可以使用`sort_values`降序排列这些数据，排序关键字通过`by`字段定义

In [91]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

PassengerId  Survived  Pclass                                   Name  \
258          259         1       1                       Ward, Miss. Anna   
679          680         1       1     Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1                 Lesurer, Mr. Gustave J   
311          312         1       1             Ryerson, Miss. Emily Borie   
742          743         1       1  Ryerson, Miss. Susan Parker "Suzette"   

        Sex   Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292              NaN        C  
679    male  36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
737    male  35.0      0      0  PC 17755  512.3292             B101        C  
311  female  18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
742  female  21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C

3. 我们来尝试在数据中添加一个新的特征（一个新的column）

In [92]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [93]:
age_categories = [age_category(age) for age in data.Age]
data['Age_category'] = age_categories

另一种实现方法是通过`.apply`函数实现
```python
data['Age_category'] = data['Age'].apply(age_category)
```

你可以通过`data.head()`命令查看新的数据的情况

In [94]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_category  
0      0         A/5 21171   7.2500   NaN        S           1.0  
1      0          PC 17599  71.2833   C85        C           2.0  
2      0  STON/O2. 3101282   7.9250   NaN        S           1.0  
3      0            113803  53.1000  C123        S           2.0  
4      0            373450   8.0500   NaN        S           2.0

### **Q2. 船上有多少男性和女性？**
- 412 男性 479 女性
- 314 男性 577 女性
- 479 男性 412 女性
- 577 男性 314 女性  $\surd$

In [95]:
# 在此编写代码回答问题
data['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

### **Q3. 当时船上有多少男性持有二级船票（Pclass=2）？**
- 104
- 108 $\surd$
- 112
- 125

In [96]:
# 在此编写代码回答问题
(data[data['Pclass']==2]['Sex']=='male').sum()

108

### **Q4. 所有乘客Fare的中位数和标准差是多少？（保留两位小数）**
- 中位数为 14.45, 标准差为 49.69 $\surd$
- 中位数为 15.1, 标准差为 12.15
- 中位数为 13.15, 标准差为 35.3
- 中位数为  17.43, 标准差为 39.1

In [97]:
# 在此编写代码回答问题
'Median of fare:''%.2f'%(data['Fare'].median()) 
'SD of Fare:''%.2f'%(data['Fare'].std())

'Median of fare:14.45'

'SD of Fare:49.69'

### **Q5. 幸存者的平均年龄是否高于最终死亡的乘客?**
- Yes
- No $\surd$

> Tips: 可以参考pandas的`groupby`命令

In [98]:
# 在此编写代码回答问题
data.groupby('Survived')['Age'].mean()
#Survivors' Av age:28.34<Victims' Av age:30.62

Survived
0    30.626179
1    28.343690
Name: Age, dtype: float64

### **Q6.年龄小于30岁的乘客是否比年龄大于60岁的乘客有更大的存活概率？ 存活概率各为多少？**
- 年轻人为22.7%，年长者为40.6% 
- 年轻人为40.6%， 年长者为22.7% $\surd$
- 年轻人为35.3%，年长者为27.4%
- 年轻人为27.4%，年长者为35.3%

> Tips: 可以查看 pandas.crosstab()函数的用法

In [99]:
# 在此编写代码回答问题
age_sur = pd.crosstab(data['Age'],data['Survived']).T
age_sur
'Number of Survivors age<30，:' '%.0f' %age_sur.loc[1,age_sur.columns<30].sum() #156
'Number of Survivors age>60，:' '%.0f'%age_sur.loc[1,age_sur.columns>60].sum() #5

Age       0.42   0.67   0.75   0.83   0.92   1.00   2.00   3.00   4.00   \
Survived                                                                  
0             0      0      0      0      0      2      7      1      3   
1             1      1      2      2      1      5      3      5      7   

Age       5.00   ...  62.00  63.00  64.00  65.00  66.00  70.00  70.50  71.00  \
Survived         ...                                                           
0             0  ...      2      0      2      3      1      2      1      2   
1             4  ...      2      2      0      0      0      0      0      0   

Age       74.00  80.00  
Survived                
0             1      0  
1             0      1  

[2 rows x 88 columns]

'Number of Survivors age<30，:156'

'Number of Survivors age>60，:5'

In [100]:
'Survival rate of Age<30:  ''%.2f' % (age_sur.loc[1,age_sur.columns<30].sum()/age_sur.loc[:,age_sur.columns<30].sum().sum())
'Survival rate of Age>60： ''%.2f' % (age_sur.loc[1,age_sur.columns>60].sum()/age_sur.loc[:,age_sur.columns>60].sum().sum())

'Survival rate of Age<30:  0.41'

'Survival rate of Age>60： 0.23'

### **Q7. 女性是否比男性有更大的存活概率? 男性与女性存活人数各占所在性别的百分之多少？**
- 男性为30.2%, 女性为46.2%
- 男性为35.7%, 女性为74.2%
- 男性为21.1%, 女性为46.2%
- 男性为18.9%, 女性为74.2% $\surd$

In [101]:
# 在此编写代码回答问题
sex_sur = data.groupby('Sex')['Survived'].value_counts()
sex_sur
sex_sur.unstack() 
sex_sur.unstack().iloc[:,1]/sex_sur.unstack().sum(axis=1)

Sex     Survived
female  1           233
        0            81
male    0           468
        1           109
Name: Survived, dtype: int64

Survived    0    1
Sex               
female     81  233
male      468  109

Sex
female    0.742038
male      0.188908
dtype: float64

### **Q8. 男性乘客中最受欢迎的名字是什么？**
- Charles
- Thomas
- William $\surd$
- John

> Tips： 请对照给出的几个选项了解名字出现的位置

In [102]:
# 在此编写代码回答问题
data_male = data[data['Sex']=='male'].reset_index(drop=True) 
names = ''
for df in data_male['Name']:
    for i in df.split(','):
        names = names+i+' '
names = names.replace('Mr. ','') #Cut Mr.
Charles_num = 0
Thomas_num = 0
William_num = 0
John_num = 0
for n in names.split(' '):
    if len(n)>3:
        if 'Charles'==n:
            Charles_num += 1
        if 'Thomas'==n:
            Thomas_num += 1
        if 'William'==n:
            William_num += 1
        if 'John'==n:
            John_num += 1 

'Charles:  ''%.f' %Charles_num 
'Thomas:   ''%.f' %Thomas_num 
'William:  ''%.f' %William_num 
'John:     ''%.f' %John_num 

'Charles:  19'

'Thomas:   16'

'William:  47'

'John:     36'

### **Q9. 男性/女性的平均年龄与`Pclass`的关系，如下声明正确的是(多选)：**
- 平均而言，一级仓中的男性年龄平均超过40岁
- 一级仓中女性的平均年龄超过40岁 
- 所有不同级别仓中男性的平均年龄都大于同一级仓中的女性 $\surd$
- 平均而言，一级仓乘客年龄大于二级票仓，二级仓大于三级票仓 $\surd$

In [103]:
# 在此编写代码回答问题
#Pclass Sex/av Age
data.groupby(['Pclass','Sex'])['Age'].mean()

Pclass  Sex   
1       female    34.611765
        male      41.281386
2       female    28.722973
        male      30.740707
3       female    21.750000
        male      26.507589
Name: Age, dtype: float64

In [104]:
#Pclass av Age
data.groupby(['Pclass'])['Age'].mean()

Pclass
1    38.233441
2    29.877630
3    25.140620
Name: Age, dtype: float64